### PR TITLE
Fix marking of saved assemblies

### LINKED ISSUE #86242: Linking an empty .NET MAUI app produces inconsistent binaries
Repro:
- Create a MAUI app with `dotnet new maui`
- Build it for MacCatalyst or iOS, eg. `dotnet build -f net8.0-maccatalyst`
- Check the linked assemblies in `obj/Debug/net8.0-maccatalyst/maccatalyst-x64/linked/`.

Observations:
- In `Microsoft.Maui.Controls.dll` the `INavigation.PopToRootAsync()` method is trimmed from the interface.
- In `Microsoft.Maui.Compatibility.dll` the `Microsoft.Maui.Controls.Compatibility.Platform.iOS.Platform` type keeps an explicit implementation of the `INavigation.PopToRootAsync()` method.

Expectations:
- Either the method is not trimmed from the interface, or all explicit implementation of the method are trimmed too.

### PR DESCRIPTION
**Pull request by sbomer**
Fixes https://github.com/dotnet/runtime/issues/86242.

This ensures the fix from https://github.com/dotnet/runtime/pull/82197 works also for assemblies which get the `save` action (which isn't supported from the command-line, but may happen in custom steps).

It also ensures that `save` assemblies get fully marked regardless of `DisableMarkingOfCopyAssemblies`.

### CODE REVIEW

File: src/tools/illink/src/linker/Linker.Steps/MarkStep.cs [hunk 1]
```diff
@@ -1410,10 +1410,14 @@ protected void MarkAssembly (AssemblyDefinition assembly, DependencyInfo reason)
...
 
-			if (ProcessReferencesStep.IsFullyPreservedAction (Annotations.GetAction (assembly))) {
-				if (!Context.TryGetCustomData ("DisableMarkingOfCopyAssemblies", out string? disableMarkingOfCopyAssembliesValue) ||
-					disableMarkingOfCopyAssembliesValue != "true")
+			var action = Annotations.GetAction (assembly);
+			if (ProcessReferencesStep.IsFullyPreservedAction (action)) {
+				if (action != AssemblyAction.Copy ||
+					!Context.TryGetCustomData ("DisableMarkingOfCopyAssemblies", out string? disableMarkingOfCopyAssembliesValue) ||
+					disableMarkingOfCopyAssembliesValue != "true") {
+					Console.Error.WriteLine($"MarkEntireAssembly {assembly.Name.Name}");
```
Comment (vitek-karas) [inline src/tools/illink/src/linker/Linker.Steps/MarkStep.cs:1418]: The `Console.Error.WriteLine` is probably a left over from debugging... should be removed.

File: src/tools/illink/src/linker/Linker.Steps/MarkStep.cs [hunk 2]
```diff
@@ -3206,7 +3210,7 @@ protected virtual void ProcessMethod (MethodDefinition method, in DependencyInfo
 			if (method.HasOverrides) {
 				var assembly = Context.Resolve (method.DeclaringType.Scope);
 				// If this method is in a Copy or CopyUsed assembly, .overrides won't get swept and we need to keep all of them
```
Comment (vitek-karas) [inline src/tools/illink/src/linker/Linker.Steps/MarkStep.cs:3212]: Also update the comment

**Review by vitek-karas — APPROVED** [review]
I wonder if there are other use cases of `Save`... this is a rather big change for those. @radekdoulik  could you please check size of Xamarin Android once this change propagates all the way there?
Also -  do you know who's tracking iOS size?

File: src/tools/illink/test/Mono.Linker.Tests.Cases/Inheritance.Interfaces/StaticInterfaceMethods/OverrideInSaveAssembly.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+{
+	[KeptMemberInAssembly ("library", typeof (IStaticAbstractMethods), "Method()", "get_Property()", "Property", "set_Property(System.Int32)")]
+    // Add a custom step which sets the assembly action of the test to "save"    
+    [SetupCompileBefore ("CustomStepSaveAssembly.dll", new[] { "Dependencies/CustomStepSaveAssembly.cs" }, new[] { "illink.dll", "Mono.Cecil.dll", "netstandard.dll" })]
+    [SetupLinkerArgument ("--custom-step", "-MarkStep:CustomStepSaveAssembly,CustomStepSaveAssembly.dll")]
+	[SetupCompileBefore ("library.dll", new[] { "Dependencies/Library.cs" })]
+	[SetupLinkerAction ("link", "library")]
+	/// <summary>
+	///	Regression test for issue: https://github.com/dotnet/runtime/issues/86242
+	///	OverridesStaticInterfaceMethods.Method() (and Property.set/get) has an entry in .overrides pointing to ISataticAbstractMethods.Method.
```
Comment (jtschuster) [inline src/tools/illink/test/Mono.Linker.Tests.Cases/Inheritance.Interfaces/StaticInterfaceMethods/OverrideInSaveAssembly.cs:23]:
```suggestion
	///	OverridesStaticInterfaceMethods.Method() (and Property.set/get) has an entry in .overrides pointing to IStaticAbstractMethods.Method.
```

**Review by jtschuster — APPROVED** [review]

**Comment by vitek-karas** [comment]
> Yes, unfortunately I don't have a great way to evaluate this. I checked link times of the repro and didn't see any significant difference with the fix.

We should not see any impact in runtime or asp.net, as those should all use full trimming everywhere.

This should only impact Xamarin and maybe Blazor scenarios.

**Pull request merged**